ES|QL Views

craigtaverner opened the pull request:
Initial support for ES|QL Views. It is based on the underlying sub-query support added in ..., and builds on the Views REST API work done in https://github.com/elastic/elasticsearch/pull/137818.

Checklist:

 - [x] Core prototype
   - [x] Extract key parts of Nik's original prototype
   - [x] Write testing infrastructure to test views within csv-spec tests (single and multi-node tests)
   - [x] Expand REST API to cover get, list, create/update, delete views 
   - [x] Add size and count limitations to stored view definitions
   - [x] Write testing infrastructure to test views in CsvTest (csv-spec unit testing)
   - [x] Expand scope to nested views
   - [x] Expand scope to multiple views and views+indexes
 - [x] Split into multiple PRs for easy review
   - [x] Multi-relation index resolution (gives CCS/CPS support)
     - [x] https://github.com/elastic/elasticsearch/pull/136780
   - [x] More resilient EsqlSpecTestCase
     - [x] https://github.com/elastic/elasticsearch/pull/136781
   - [x] View CRUD with REST
     - [x] https://github.com/elastic/elasticsearch/pull/137818
   - [x] Rebase prototype on the above to get cleaner, simple feature PR
 - [ ] Enhancements
   - [x] Prototype CCS and CPS
   - [x] Enforce view/index name uniqueness (done in separate PR)
   - [x] Add serialization tests to test source deserialization in re-written plans
   - [x] Ensure views are gated behind SNAPSHOT and/or feature flag
   - [ ] Increate test coverage at least to the same extent as sub-queries
   - [ ] Control and/or restrict nesting/branching depth/width

craigtaverner committed:
ESQL: Views

Initial work on getting views integrated with ES|QL
* Access views within FROM command
* Add csv-spec tests for non-branching and branching views
* Get CsvTests working on views
* Updated command-line CsvTestDataLoader for views, and multiple runs
* Disable MultiClusterSpecIT for Views (test rewrites "FROM name" to "FROM *:name", which is not yet supported)
  * Also make sure we do not even load views for these tests, because that messes with ENRICH and LOOKUP JOIN tests
* Ensure views capabilities are tied to the views feature flag
* Views with SORT also must have explicit LIMIT (Due to PR#139058)

idegtiarenko commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlCCSUtils.java:
```diff
@@ -304,24 +337,29 @@ static void updateExecutionInfoAtEndOfPlanning(EsqlExecutionInfo execInfo) {
+                    System.arraycopy(a, 0, combined, 0, a.length);
+                    System.arraycopy(b, 0, combined, a.length, b.length);
+                    return combined;
+                })
```
```suggestion
                .toArray(String[]::new)
```

idegtiarenko commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlCCSUtils.java:
```diff
@@ -304,24 +337,29 @@ static void updateExecutionInfoAtEndOfPlanning(EsqlExecutionInfo execInfo) {
-                false
-            );
+            String[] indexExpressions = indexPatterns.stream()
+                .map(indexPattern -> Strings.splitStringByCommaToArray(indexPattern.indexPattern()))
```
This seems like it is combining all patterns into one. Unfortunately this is not safe to do.
For example:
* logs-*,-*2025
* metrics-*
If above 2 patterns are combined, metrics-2025 will be unintentionally excluded from the final result set.
Here this result is used to initialize CC state. If metrics-2025 is the only index on `r1` it will not be initialized. This would lead to runtime error when populating the cluster info when actually querying `r1`.

I am currently working to avoid the entire `initCrossClusterState`, so likely this will no longer be needed.
See https://github.com/elastic/elasticsearch/pull/136804

craigtaverner replied:
OK. This problem seems to be a potential bug for subqueries and views in CCS, but should not cause any problems for the normal single FROM command case. We can merge this early, and fix the problem properly with #136804.

idegtiarenko commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/PreAnalyzer.java:
```diff
@@ -43,17 +44,19 @@ public PreAnalysis preAnalyze(LogicalPlan plan) {
+                IndexMode m2 = indexes.get(p.indexPattern());
+                throw new IllegalStateException(
+                    "index pattern '" + p.indexPattern() + "' found with with different index mode: " + m2 + " != " + m1
+                );
```
I believe IndexMode is read from the index settings. Is it even possible to find the same pattern in 2 different index modes?

craigtaverner replied:
OK. Let's try use an assert.

idegtiarenko commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -532,21 +547,19 @@ private void resolveIndicesAndAnalyze(
-            }
-            return r;
-        })
+            l -> preAnalyzeMainIndices(preAnalysis.indexes().keySet().iterator(), preAnalysis, executionInfo, result, requestFilter, l)
```
Iterating over `entrySet()` could avoid additional lookup later on for the index mode.

idegtiarenko commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -532,21 +547,19 @@ private void resolveIndicesAndAnalyze(
-                LOGGER.debug("No more clusters to search, ending analysis stage");
-                throw new NoClustersToSearchException();
-            }
-            return r;
```
For my understanding, does this check causes any troubles with view resolution?

idegtiarenko commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/view/InMemoryViewService.java:
```diff
@@ -0,0 +1,48 @@
+ * Simple implementation of {@link ClusterViewService} that keeps the views in memory.
+ * This is useful for testing.
+ */
+public class InMemoryViewService extends ViewService {
```
This could be in test sources.

NIT:
It would be nice to avoid abstract part and only have cluster state based implementation in production code for simplicity with in-memory version in test overriding specific lookup parts.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/view/ViewService.java:
```diff
@@ -0,0 +1,211 @@
+                List<String> indexes = new ArrayList<>();
+                List<LogicalPlan> subqueries = new ArrayList<>();
+                for (String name : ur.indexPattern().indexPattern().split(",")) {
+                    name = name.trim();
```
I think this is still prototype code. I want to improve it when we create the views CRUD PR.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/view/ClusterViewService.java:
```diff
@@ -0,0 +1,71 @@
+    @SuppressForbidden(reason = "legacy usage of unbatched task") // TODO add support for batching here
+    private void submitUnbatchedTask(@SuppressWarnings("SameParameterValue") String source, ClusterStateUpdateTask task) {
+        clusterService.submitUnbatchedStateUpdateTask(source, task);
+    }
```
Let's do this as a followup

elasticsearchmachine commented:
Hi @craigtaverner, I've created a changelog YAML for you.

craigtaverner committed:
Split ViewResolver out of ViewService and support wildcards

craigtaverner committed:
Improve views resolution logging and disabled FROM * with views tests

craigtaverner commented on x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/EsqlSpecTestCase.java:
```diff
@@ -245,6 +264,25 @@ protected void ensureTestClustersAreOk(Exception failure) {
         }
     }
 
+    private static final Set<String> FROM_STAR_TESTS = Set.of(
```
I thought about that, but also thought that it was useful to see that as many existing tests as possible pass unchanged and are unaffected by the existence of views. I like the fact that the existing queries that are affected by the existence of views are a small subset, and a subset I'd like to keep explicit so we keep this consequence in mind. This is then also similar to the enrich policies and lookup indices, which exist even for queries that make no use of them.

craigtaverner committed:
Capture view name in Subquery for logging and debugging

craigtaverner committed:
Use Regex.simpleMatch for wildcards

julian-elastic reviewed (approved):
Overall looks good, great work! I left some comments about additional testing/questions to resolve before we check it in.

julian-elastic commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/view/ViewResolver.java:
```diff
@@ -0,0 +1,313 @@
+        LogicalPlan rewritten = plan.transformDown(LogicalPlan.class, p -> {
+            switch (p) {
+                case UnionAll union -> {
+                    // UnionAll is the result of this re-writing, so we assume rewriting is completed
```
Alternatively, can't we just check if the children of the UnionAll are named subqueries to know if we created it or not?

craigtaverner commented on x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/EsqlSpecTestCase.java:
```diff
@@ -245,6 +264,25 @@ protected void ensureTestClustersAreOk(Exception failure) {
+
+    private static final Set<String> AIRPORTS_STAR_TESTS = Set.of("LucenePushdownMixOrAnd", "LucenePushdownMultipleIndices");
+
+    // Some tests will fail if views are defined (notably tests with `FROM *` which are non-deterministic)
```
There are existing tests that started failing, and a new test I added called "views.QueryThatWillNotBeDeterministicWithViews" to have a minimal reproduction of this. However, this is related to the work on implicit LIMIT in fork, and @fang-xing-esql removed that, so the behaviour has changed. I will verify if all these tests still need to be blocked after the latest changes to main.

craigtaverner committed:
Reduce list of tests to ignore from views setup

After changes in main, it seems more tests survive the presence of views

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into views_prototype

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/view/ViewResolver.java:
```diff
@@ -0,0 +1,277 @@
+                seenWildcards.add(name);
+                // If the name includes a wildcard, expand it to all matching views,
+                // leaving the original wildcard name in place for index resolution
+                String namePattern = name.replace("*", ".*");
```
Done

craigtaverner committed:
Remove a few tests using `FROM *`

craigtaverner committed:
Added tests that show warnings get propagated through views

Note that the line-numbers are relative to the view query, and this will likely change once we fix source errors.

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into views_prototype

craigtaverner committed:
Only load views containing multi-values for tests that assert on the warnings

craigtaverner committed:
Some fixes to MV-views loading

craigtaverner committed:
Simplify view loading for tests (only load for tests that need views)

craigtaverner committed:
Support view name in Source and Configuration so errors and warnings are correct

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into views_prototype

craigtaverner committed:
Updated transport version after merging main

craigtaverner committed:
Updated transport version after merging main

julian-elastic reviewed (approved):
Overall, still looks good with the latest changes. However, I would recommend fixing tech debt where passing something from source to warnings requires 400+ files change.

craigtaverner committed:
Rather pass source to createWarnings, so we can simplify generated code

This also reduces future refactoring

craigtaverner committed:
Remove older methods from Warnings, simplify and update tests

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into views_prototype

craigtaverner committed:
Merge branch 'main' into views_prototype

craigtaverner committed:
Merge branch 'main' into views_prototype

craigtaverner committed:
Create ViewResolver.NOOP for release builds

craigtaverner committed:
Registered ViewResolver.NOOP did not work, so we go for public noop constructor instead.

craigtaverner committed:
Cleaned up view settings based on code review

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into views_prototype

craigtaverner committed:
Missing generated files

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into views_prototype

craigtaverner committed:
Regenerated transport version

craigtaverner committed:
Added yaml tests for views (including negative tests)

craigtaverner committed:
Different error messages on serverless

elasticsearchmachine commented:
@craigtaverner according to this PR's labels, I need to update the changelog YAML, but I can't because the PR is closed. Please either update the changelog yourself on the appropriate branch, or adjust the labels. Specifically:

   * The PR is not labelled `release highlight` but the changelog has a `highlight` section
